// @cerebras/cerebras_cloud_sdk/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var _Cerebras_instances, _a, _Cerebras_baseURLOverridden;
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnprocessableEntityError = exports.PermissionDeniedError = exports.InternalServerError = exports.AuthenticationError = exports.BadRequestError = exports.RateLimitError = exports.ConflictError = exports.NotFoundError = exports.APIUserAbortError = exports.APIConnectionTimeoutError = exports.APIConnectionError = exports.APIError = exports.CerebrasError = exports.fileFromPath = exports.toFile = exports.Cerebras = void 0;
const Core = __importStar(require("./core.js"));
const Errors = __importStar(require("./error.js"));
const Uploads = __importStar(require("./uploads.js"));
const API = __importStar(require("./resources/index.js"));
const completions_1 = require("./resources/completions.js");
const models_1 = require("./resources/models.js");
const chat_1 = require("./resources/chat/chat.js");
/**
 * API Client for interfacing with the Cerebras API.
 */
class Cerebras extends Core.APIClient {
    /**
     * API Client for interfacing with the Cerebras API.
     *
     * @param {string | undefined} [opts.apiKey=process.env['CEREBRAS_API_KEY'] ?? undefined]
     * @param {string} [opts.baseURL=process.env['CEREBRAS_BASE_URL'] ?? https://api.cerebras.ai] - Override the default base URL for the API.
     * @param {number} [opts.timeout=1 minute] - The maximum amount of time (in milliseconds) the client will wait for a response before timing out.
     * @param {number} [opts.httpAgent] - An HTTP agent used to manage HTTP(s) connections.
     * @param {Core.Fetch} [opts.fetch] - Specify a custom `fetch` function implementation.
     * @param {number} [opts.maxRetries=2] - The maximum number of times the client will retry a request.
     * @param {Core.Headers} opts.defaultHeaders - Default headers to include with every request to the API.
     * @param {Core.DefaultQuery} opts.defaultQuery - Default query parameters to include with every request to the API.
     * @param {boolean | undefined} opts.warmTCPConnection - Whether to warm TCP connection in the constructor.
     */
    constructor({ baseURL = Core.readEnv('CEREBRAS_BASE_URL'), apiKey = Core.readEnv('CEREBRAS_API_KEY'), warmTCPConnection = true, ...opts } = {}) {
        if (apiKey === undefined) {
            throw new Errors.CerebrasError("The CEREBRAS_API_KEY environment variable is missing or empty; either provide it, or instantiate the Cerebras client with an apiKey option, like new Cerebras({ apiKey: 'My API Key' }).");
        }
        const options = {
            apiKey,
            ...opts,
            baseURL: baseURL || `https://api.cerebras.ai`,
        };
        super({
            baseURL: options.baseURL,
            baseURLOverridden: baseURL ? baseURL !== 'https://api.cerebras.ai' : false,
            timeout: options.timeout ?? 60000 /* 1 minute */,
            httpAgent: options.httpAgent,
            maxRetries: options.maxRetries,
            fetch: options.fetch,
        });
        _Cerebras_instances.add(this);
        this.chat = new API.Chat(this);
        this.completions = new API.Completions(this);
        this.models = new API.Models(this);
        this._options = options;
        this.apiKey = apiKey;
        if (warmTCPConnection) {
            // Since this runs async, it's possible for DEBUG messages to
            // be printed after test ends, which will cause a warming.
            //
            // Doesn't seem to be an easy way to block until this promise is fulfilled.
            (async () => {
                try {
                    await this.get('/v1/tcp_warming', {
                        timeout: 1000,
                        maxRetries: 0,
                    });
                }
                catch (e) {
                    Core.debug(`TCP Warming had exception: ${e}`);
                }
            })();
        }
    }
    defaultQuery() {
        return this._options.defaultQuery;
    }
    defaultHeaders(opts) {
        return {
            ...super.defaultHeaders(opts),
            ...this._options.defaultHeaders,
        };
    }
    authHeaders(opts) {
        return { Authorization: `Bearer ${this.apiKey}` };
    }
}
exports.Cerebras = Cerebras;
_a = Cerebras, _Cerebras_instances = new WeakSet(), _Cerebras_baseURLOverridden = function _Cerebras_baseURLOverridden() {
    return this.baseURL !== 'https://api.cerebras.ai';
};
Cerebras.Cerebras = _a;
Cerebras.DEFAULT_TIMEOUT = 60000; // 1 minute
Cerebras.CerebrasError = Errors.CerebrasError;
Cerebras.APIError = Errors.APIError;
Cerebras.APIConnectionError = Errors.APIConnectionError;
Cerebras.APIConnectionTimeoutError = Errors.APIConnectionTimeoutError;
Cerebras.APIUserAbortError = Errors.APIUserAbortError;
Cerebras.NotFoundError = Errors.NotFoundError;
Cerebras.ConflictError = Errors.ConflictError;
Cerebras.RateLimitError = Errors.RateLimitError;
Cerebras.BadRequestError = Errors.BadRequestError;
Cerebras.AuthenticationError = Errors.AuthenticationError;
Cerebras.InternalServerError = Errors.InternalServerError;
Cerebras.PermissionDeniedError = Errors.PermissionDeniedError;
Cerebras.UnprocessableEntityError = Errors.UnprocessableEntityError;
Cerebras.toFile = Uploads.toFile;
Cerebras.fileFromPath = Uploads.fileFromPath;
Cerebras.Chat = chat_1.Chat;
Cerebras.Completions = completions_1.Completions;
Cerebras.Models = models_1.Models;
var uploads_1 = require("./uploads.js");
Object.defineProperty(exports, "toFile", { enumerable: true, get: function () { return uploads_1.toFile; } });
Object.defineProperty(exports, "fileFromPath", { enumerable: true, get: function () { return uploads_1.fileFromPath; } });
var error_1 = require("./error.js");
Object.defineProperty(exports, "CerebrasError", { enumerable: true, get: function () { return error_1.CerebrasError; } });
Object.defineProperty(exports, "APIError", { enumerable: true, get: function () { return error_1.APIError; } });
Object.defineProperty(exports, "APIConnectionError", { enumerable: true, get: function () { return error_1.APIConnectionError; } });
Object.defineProperty(exports, "APIConnectionTimeoutError", { enumerable: true, get: function () { return error_1.APIConnectionTimeoutError; } });
Object.defineProperty(exports, "APIUserAbortError", { enumerable: true, get: function () { return error_1.APIUserAbortError; } });
Object.defineProperty(exports, "NotFoundError", { enumerable: true, get: function () { return error_1.NotFoundError; } });
Object.defineProperty(exports, "ConflictError", { enumerable: true, get: function () { return error_1.ConflictError; } });
Object.defineProperty(exports, "RateLimitError", { enumerable: true, get: function () { return error_1.RateLimitError; } });
Object.defineProperty(exports, "BadRequestError", { enumerable: true, get: function () { return error_1.BadRequestError; } });
Object.defineProperty(exports, "AuthenticationError", { enumerable: true, get: function () { return error_1.AuthenticationError; } });
Object.defineProperty(exports, "InternalServerError", { enumerable: true, get: function () { return error_1.InternalServerError; } });
Object.defineProperty(exports, "PermissionDeniedError", { enumerable: true, get: function () { return error_1.PermissionDeniedError; } });
Object.defineProperty(exports, "UnprocessableEntityError", { enumerable: true, get: function () { return error_1.UnprocessableEntityError; } });
exports = module.exports = Cerebras;
exports.default = Cerebras;
